### DISCUSSION TITLE
Docs: explain how to deploy HTTPS with own certificate

### DISCUSSION THREAD

**Discussion by pawamoy** [fastapi/fastapi#7843] [Questions] [answered]
### Is your feature request related to a problem

A docs problem, yes 😛 

I want to use HTTPS for my API. The docs suggest to use a Docker Swarm with Traefik and Let's Encrypt. I can't (it's for work). We use a Kubernetes cluster and our own SSL certificates. Then there is a link to the FastAPI project generator, but there are a *lot* of files in there. I'm not sure to even understand which of Gunicorn, Uvicorn or FastAPI (or Starlette?) is responsible for the HTTPS part.

So the question, in the end, is: how do I deploy a FastAPI app with HTTPS enabled, using an SSL certificate I already have?

### The solution you would like
As asked above, I would like another section to be added to the Deployment / HTTPS part of the docs explaning how to deploy a FastAPI API with HTTPS enabled, with my own certificates 🙂 

---

EDIT: I think I got it. It should be configured [in Gunicorn](http://docs.gunicorn.org/en/latest/settings.html#ssl) (if used to manage Uvicorn), or [in Uvicorn](https://www.uvicorn.org/settings/#https) itself.

**Comment by euri10** [comment]
> Then there is a link to the FastAPI project generator, but there are a _lot_ of files in there. I'm not sure to even understand which of Gunicorn, Uvicorn or FastAPI (or Starlette?) is responsible for the HTTPS part.

none of those it happens in the traefik container which is a reverse-proxy
you have many possibilities to setup ssl, most of the time people do it on the reverse proxy , but as you pointed out in your links you can do it on the webserver directly

**Comment by prostomarkeloff** [comment, marked as answer]
> ### Is your feature request related to a problem
> 
> A docs problem, yes stuck_out_tongue
> 
> I want to use HTTPS for my API. The docs suggest to use a Docker Swarm with Traefik and Let's Encrypt. I can't (it's for work). We use a Kubernetes cluster and our own SSL certificates. Then there is a link to the FastAPI project generator, but there are a _lot_ of files in there. I'm not sure to even understand which of Gunicorn, Uvicorn or FastAPI (or Starlette?) is responsible for the HTTPS part.
> 
> So the question, in the end, is: how do I deploy a FastAPI app with HTTPS enabled, using an SSL certificate I already have?
> ### The solution you would like
> 
> As asked above, I would like another section to be added to the Deployment / HTTPS part of the docs explaning how to deploy a FastAPI API with HTTPS enabled, with my own certificates slightly_smiling_face
> 
> EDIT: I think I got it. It should be configured [in Gunicorn](http://docs.gunicorn.org/en/latest/settings.html#ssl) (if used to manage Uvicorn), or [in Uvicorn](https://www.uvicorn.org/settings/#https) itself.

Heyo! Ye, u can do it.
Uvicorn [supports it](https://www.uvicorn.org/deployment/). You also can check the gunicorn docs.

**Comment by pawamoy** [comment]
Yeah I usually setup SSL in NginX configuration, but in this case, I read from the author that NginX is not really needed (unless you need to serve static files) since Uvicorn is production ready.

I think the documentation should mention that.

**Comment by tiangolo** [comment]
@pawamoy actually you can use Nginx and setup TLS in there as you normally do. And then use a proxypass to Uvicorn. You don't *need* Nginx, but you can use it if you want. Especially if you are already familiar with it.

**Comment by borland667** [comment]
@pawamoy You can use you the nginx ingress controller and certmanager with k8s and have a standard deployment, service and ingress configuration for your fastapi project.

**Comment by tiangolo** [comment]
Does that solve your problem @pawamoy ? May we close this issue?

**Comment by pawamoy** [comment]
I think we can close yes.
